File: src/decoder.rs
use crate::{
    boardmanager::BoardStack,
    cache::CacheEntryKey,
    dataformat::{ZeroEvaluationAbs, ZeroValuesPov},
    debug_print,
    mcts_trainer::{Net, Node, Tree, Wdl},
    mvs::get_contents,
};
use cozy_chess::{Color, Move, Piece, Rank, Square};
use lru::LruCache;
use tch::{Device, IValue, Kind, Tensor};

/// evaluates the current state of the chess board
/// returns a tuple of `Tensors` - in the order (value, policy)
pub fn eval_state(board: Tensor, net: &Net) -> anyhow::Result<(Tensor, Tensor)> {
    let b = board.reshape([-1, 21, 8, 8]);
    let b: Tensor = b.to(net.device);
    let board = IValue::Tensor(b);

    let output = net.net.forward_is(&[board])?;

    let output_tensor = match output {
        IValue::Tuple(b) => b,
        a => panic!("the output is not an IValue {:?}", a),
    };

    let (board_eval, policy) = (&output_tensor[0], &output_tensor[1]);
    let board_eval = match board_eval {
        IValue::Tensor(b) => b,
        a => panic!("the output is not a Tensor {:?}", a),
    };
    let policy = match policy {
        IValue::Tensor(b) => b,
        a => panic!("the output is not a Tensor {:?}", a),
    };

    // Move the results to CPU to reduce GPU memory usage
    let board_eval = board_eval.to(Device::Cpu);
    let policy = policy.to(Device::Cpu);
    // Drop the original GPU tensors to free memory
    drop(output_tensor);

    Ok((board_eval, policy))
}

/// extracts scalar and piece square data from the board state. used to convert board to use as nn inputs
/// to save compute, the boolean-representable data is returned as a single Vec<bool> before broadcasting into a single input `Tensor`
/// returns 2 items, a Vec<f32> for the board inputs and Vec<bool>

// FULL LIST OF INFORMATION NEEDED FOR NN INPUTS:

/// sq1 - white's turn
/// sq2 - black's turn
/// sq3, sq4 - castling pos l + r (us)
/// sq5, sq6 - castling pos l + r (opponent)
/// sql7, sql8 -  sqs for bits for the repetition counter
/// sq9 - sq20 - sqs for turn to move + non-turn to move's pieces
/// sq21 - en passant square if any

pub fn board_data(bs: &BoardStack) -> (Vec<f32>, Vec<bool>) {
    let us = bs.board().side_to_move();
    let mut scalar_data = vec![0.0; 8];
    if bs.board().side_to_move() == Color::Black {
        scalar_data[1] = 1.0;
    } else {
        scalar_data[0] = 1.0;
    }

    let li: [Color; 2] = if us == Color::White {
        // can't use Colour::ALL since the order of Colour::ALL is always going to be [white, black]
        [Color::White, Color::Black]
    } else {
        [Color::Black, Color::White]
    };

    let mut c = 2;

    for color in li {
        let l_rights = bs.board().castle_rights(color).long;
        let s_rights = bs.board().castle_rights(color).short;

        scalar_data[c] = if l_rights.is_some() { 1.0 } else { 0.0 };
        scalar_data[c + 1] = if s_rights.is_some() { 1.0 } else { 0.0 };
        c += 2
    }

    scalar_data[6] = bs.get_reps() as f32;
    scalar_data[7] = bs.board().halfmove_clock() as f32;

    let mut counter = 0;
    let mut pieces_sqs: Vec<bool> = vec![false; 64 * 12];
    for colour in li {
        for piece in Piece::ALL {
            for tile in bs.board().colored_pieces(colour, piece) {
                if li[0] == Color::Black {
                    pieces_sqs[(63 - (tile.rank() as usize * 8 + (7 - tile.file() as usize)))
                        + (64 * counter)] = true;
                } else {
                    pieces_sqs[(tile.rank() as usize * 8 + tile.file() as usize) + 64 * counter] =
                        true;
                }
            }
            counter += 1
        }
    }

    let is_ep = bs.board().en_passant();
    debug_print!("    board FEN: {}", bs.board());
    debug_print!("En passant status: {:?}", is_ep);

    let mut sq21: Vec<bool> = vec![false; 64];
    if let Some(is_ep) = is_ep {
        if us == Color::White {
            // 4 for white and 5 for black for victim
            let row = Rank::Fourth;
            let ep_sq = Square::new(is_ep, row);
            sq21[ep_sq.rank() as usize * 8 + ep_sq.file() as usize] = true;
        } else {
            let row = Rank::Fifth;
            let ep_sq = Square::new(is_ep, row);
            sq21[63 - (ep_sq.rank() as usize * 8 + (7 - ep_sq.file() as usize))] = true;
        }
    };
    pieces_sqs.extend(sq21);
    (scalar_data, pieces_sqs)
}

/// converts the board state into a representation used for nn inputs
pub fn convert_board(bs: &BoardStack) -> Tensor {
    let mut all_data: Vec<f32> = Vec::new();

    let (scalar_data, pieces_sqs) = board_data(bs);

    for v in &scalar_data {
        for _ in 0..64 {
            all_data.push(*v);
        }
    }

    all_data.extend(pieces_sqs.iter().map(|&x| x as u8 as f32));

    Tensor::from_slice(&all_data) // all_data is 1d
}

/// extracts the policy indices from `mvs.rs` given board state in order to filter raw output from neural network
pub fn extract_policy(bs: &BoardStack, contents: &'static [Move]) -> (Vec<Move>, Vec<usize>) {
    let mut legal_moves: Vec<Move> = Vec::new();
    bs.board().generate_moves(|moves| {
        // Unpack dense move set into move list
        legal_moves.extend(moves);
        false
    });

    let mut fm: Vec<Move> = Vec::new();
    if bs.board().side_to_move() == Color::Black {
        // flip move
        for mv in &legal_moves {
            fm.push(Move {
                from: mv.from.flip_rank(),
                to: mv.to.flip_rank(),
                promotion: mv.promotion,
            })
        }
    } else {
        fm.clone_from(&legal_moves);
    }

    legal_moves = fm;

    let mut idx_li: Vec<usize> = Vec::new();

    for mov in &legal_moves {
        if let Some(idx) = contents.iter().position(|x| mov == x) {
            idx_li.push(idx);
        }
    }

    (legal_moves, idx_li)
}

/// processes the board output after nn evaluation and updates the tree with new nodes
pub fn process_board_output(
    output: (&Tensor, &Tensor),
    selected_node_idx: &usize,
    tree: &mut Tree,
    bs: &BoardStack,
    cache: &mut LruCache<CacheEntryKey, ZeroEvaluationAbs>,
) -> Vec<usize> {
    let contents = get_contents();
    let (board_eval, policy) = output;
    let board_eval = board_eval.squeeze();
    let board_evals: Vec<f32> = Vec::try_from(board_eval).expect("Error");

    let value: f32 = board_evals[0].tanh();

    let wdl_logits: Tensor = Tensor::from_slice(&board_evals[1..4]);
    let wdl = Tensor::softmax(&wdl_logits, 0, Kind::Float);

    let wdl: Vec<f32> = Vec::try_from(wdl).expect("Error");
    let wdl = Wdl {
        w: wdl[0],
        d: wdl[1],
        l: wdl[2],
    };
    let moves_left = board_evals[4];

    let policy = policy.squeeze();
    let policy: Vec<f32> = Vec::try_from(policy).expect("Error");
    let value = f32::try_from(value).expect("Error");

    // step 1 - get the corresponding idx for legal moves

    let (legal_moves, idx_li) = extract_policy(bs, contents);

    // step 2 - using the idx in step 1, index all the policies involved

    let mut pol_list: Vec<f32> = Vec::new();
    for id in &idx_li {
        pol_list.push(policy[*id]);
    }

    // step 3 - apply softmax to policy

    let sm = Tensor::from_slice(&pol_list);

    let sm = Tensor::softmax(&sm, 0, Kind::Float);

    let pol_list: Vec<f32> = Vec::try_from(sm).expect("Error");

    // step 4 - iteratively append nodes into class
    let mut counter = 0;

    // ensure that moves_left is positive - there are no negative moves
    let moves_left = 0.0_f32.max(moves_left);
    let selected_node_net_evaluation = ZeroValuesPov {
        value,
        wdl,
        moves_left,
    }
    .to_absolute(bs.board().side_to_move());

    tree.nodes[*selected_node_idx].net_evaluation = selected_node_net_evaluation;
    let ct = tree.nodes.len();
    for (mv, pol) in legal_moves.iter().zip(pol_list.iter()) {
        let fm: Move;
        // flip move if the current move is a black move
        if bs.board().side_to_move() == Color::Black {
            fm = Move {
                from: mv.from.flip_rank(),
                to: mv.to.flip_rank(),
                promotion: mv.promotion,
            };
        } else {
            fm = *mv;
        }
        let child = Node::new(*pol, Some(*selected_node_idx), Some(fm));

        tree.nodes.push(child);

        counter += 1
    }
    tree.nodes[*selected_node_idx].children = ct..ct + counter;

    cache.put(
        CacheEntryKey {
            hash: bs.board().hash(),
            halfmove_clock: bs.board().halfmove_clock(),
        },
        ZeroEvaluationAbs {
            values: selected_node_net_evaluation,
            policy: pol_list,
        },
    );

    idx_li
}

File: src/python/decoder.py
"""Python port of the Rust network input/output decoding (`src/decoder.rs`).

This module converts a python-chess board into the 21-plane canonical network
input tensor, and decodes raw network outputs (value, policy) into per-move
probabilities over the legal moves.

Coordinate / perspective contract (see ENGINE_SEMANTIC_SPEC.md):
- Piece planes: first 6 are the side to move ("us"), next 6 the opponent,
  each ordered pawn, knight, bishop, rook, queen, king.
- If Black is to move, every board square is rotated 180 degrees
  (flat index i -> 63 - i); White to move keeps physical squares.
- The en-passant plane uses the same square rotation.
- Castling scalar channels are ordered us (long, short) then opponent (long,
  short). Channel 0/1 are white-to-move / black-to-move indicators, channel 6
  is the repetition count, channel 7 the halfmove clock.
- Policy indices live in the fixed White-absolute 1880-move space defined by
  `src/mvs.rs`. Black moves are rank-flipped (mirror, not rotation) before
  lookup, exactly like the Rust `extract_policy`.
"""

import math
import os

import chess
import torch

# ---------------------------------------------------------------------------
# Move table (mirror `src/mvs.rs` CONTENTS_COZY exactly, via `list.txt`)
# ---------------------------------------------------------------------------


def _load_move_list():
    """Load the 1880-entry White-absolute move list from `list.txt`.

    `list.txt` holds the exact ordered move strings (UCI) matching
    `src/mvs.rs` CONTENTS_COZY, one per line. Promotions carry a trailing
    piece letter (q/r/b/n).
    """
    path = os.path.join(os.path.dirname(os.path.abspath(__file__)), "list.txt")
    with open(path, encoding="utf-8") as f:
        moves = [chess.Move.from_uci(line.strip()) for line in f if line.strip()]
    return moves


CONTENTS = _load_move_list()
assert len(CONTENTS) == 1880, f"expected 1880 moves, got {len(CONTENTS)}"
assert len(set(CONTENTS)) == 1880, "duplicate moves in table"

# move -> flat policy index (mirrors the Rust `contents.iter().position` lookup)
MOVE_TO_INDEX = {mv: i for i, mv in enumerate(CONTENTS)}

# piece plane order shared by both sides: pawn .. king (matches cozy-chess Piece::ALL)
PIECE_ORDER = [
    chess.PAWN,
    chess.KNIGHT,
    chess.BISHOP,
    chess.ROOK,
    chess.QUEEN,
    chess.KING,
]


def _flip_rank(square):
    """Mirror a square vertically (rank flip), used for Black policy lookup."""
    return chess.square(chess.square_file(square), 7 - chess.square_rank(square))


# ---------------------------------------------------------------------------
# Board -> tensor  (port of Rust `board_data` + `convert_board`)
# ---------------------------------------------------------------------------


def convert_board(board, bigl=None):
    """Convert a python-chess `Board` into the canonical 21x8x8 input tensor.

    Returns `(tensor, bigl)`. `bigl` is kept only for backward compatibility
    with the old API and is always an empty tensor.
    """
    us = board.turn
    opponent = not us

    # --- scalar channels -------------------------------------------------
    scalars = [0.0] * 8
    if us == chess.WHITE:
        scalars[0] = 1.0
    else:
        scalars[1] = 1.0

    # castling rights, us first then opponent (long then short)
    scalars[2] = 1.0 if board.has_queenside_castling_rights(us) else 0.0
    scalars[3] = 1.0 if board.has_kingside_castling_rights(us) else 0.0
    scalars[4] = 1.0 if board.has_queenside_castling_rights(opponent) else 0.0
    scalars[5] = 1.0 if board.has_kingside_castling_rights(opponent) else 0.0

    # repetition count of the current position + halfmove clock
    scalars[6] = float(_repetition_count(board))
    scalars[7] = float(board.halfmove_clock)

    # --- piece planes ----------------------------------------------------
    planes = torch.zeros(13, 8, 8)
    counter = 0
    for color in (us, opponent):
        for piece in PIECE_ORDER:
            for sq in board.pieces(piece, color):
                if us == chess.BLACK:
                    sq = 63 - sq  # 180-degree rotation for Black to move
                planes[counter, sq // 8, sq % 8] = 1.0
            counter += 1

    # --- en-passant plane ------------------------------------------------
    # Mirrors Rust: White to move marks the victim square on Rank::Fourth
    # (rank index 3); Black to move marks Rank::Fifth (rank index 4) and
    # applies the same 180-degree rotation as the piece planes.
    if board.ep_square is not None:
        if us == chess.WHITE:
            ep_sq = chess.square(chess.square_file(board.ep_square), 3)
        else:
            ep_sq = 63 - chess.square(chess.square_file(board.ep_square), 4)
        planes[12, ep_sq // 8, ep_sq % 8] = 1.0

    # scalars are broadcast across the board and placed before the bool planes
    scalar_planes = (
        torch.tensor(scalars, dtype=torch.float32).view(8, 1, 1).expand(8, 8, 8)
    )
    tensor = torch.cat([scalar_planes, planes], dim=0)  # (21, 8, 8)

    if bigl is None:
        bigl = torch.tensor([])
    return tensor, bigl


def _repetition_count(board):
    """Number of times the current position occurred before (Rust `get_reps`).

    Uses python-chess's Zobrist transposition key, matching the Rust
    `BoardStack::get_reps` which hashes prior positions on the move stack.
    Pops and re-pushes from a snapshot of the move stack so the board is left
    unmodified.
    """
    target = board._transposition_key()
    stack = list(board.move_stack)
    count = 0
    for _ in range(len(stack)):
        board.pop()
        if board._transposition_key() == target:
            count += 1
    for move in stack:
        board.push(move)
    return count


# ---------------------------------------------------------------------------
# Network output -> legal policy  (port of Rust `extract_policy` + decode)
# ---------------------------------------------------------------------------


def extract_policy(board):
    """Return `(legal_moves, idx_li)` for the position.

    `legal_moves` are python-chess moves canonicalised into the White-absolute
    policy space (rank-flipped for Black to move); `idx_li` are their flat
    policy indices in the same order as the board's legal-move generator.
    """
    legal_moves = []
    for mv in board.legal_moves:
        if board.turn == chess.BLACK:
            mv = chess.Move(
                _flip_rank(mv.from_square),
                _flip_rank(mv.to_square),
                promotion=mv.promotion,
            )
        legal_moves.append(mv)

    idx_li = [MOVE_TO_INDEX[mv] for mv in legal_moves]
    return legal_moves, idx_li


def decode_nn_output(value, policy, board):
    """Decode raw network outputs for a position.

    `value` is the network's 5-element value head output
    `(scalar logit, wdl logit x3, moves-left)` and `policy` the 1880 logits.
    Both may be torch tensors or anything convertible; batched inputs are
    squeezed.

    Returns `(value, wdl, moves_left, legal_moves, legal_lookup, policy_list,
    best_move)`:
      - `value`: scalar in [-1, 1] (tanh of the raw logit), from the side to
        move's perspective;
      - `wdl`: `[win, draw, loss]` probabilities for the side to move;
      - `moves_left`: clamped to be non-negative;
      - `legal_moves`: canonicalised legal moves (White-absolute space);
      - `legal_lookup`: `{uci_string: softmax_probability}` for legal moves;
      - `policy_list`: probabilities aligned with `legal_moves`;
      - `best_move`: UCI string of the highest-probability legal move
        (physical coordinates for the side to move).
    """
    if not torch.is_tensor(value):
        value = torch.tensor(value, dtype=torch.float32)
    if not torch.is_tensor(policy):
        policy = torch.tensor(policy, dtype=torch.float32)

    value = value.squeeze().float().cpu()
    policy = policy.squeeze().float().cpu()

    raw = value.tolist()
    scalar = math.tanh(raw[0])

    wdl_logits = torch.tensor(raw[1:4], dtype=torch.float32)
    wdl = torch.softmax(wdl_logits, dim=0).tolist()

    moves_left = max(0.0, raw[4])

    legal_moves, idx_li = extract_policy(board)
    legal_logits = policy[idx_li]
    policy_probs = torch.softmax(legal_logits, dim=0).tolist()

    legal_lookup = {mv.uci(): p for mv, p in zip(legal_moves, policy_probs)}

    # best move in physical coordinates for the side to move
    best_canonical = legal_moves[
        max(range(len(legal_moves)), key=lambda i: policy_probs[i])
    ]
    if board.turn == chess.BLACK:
        best_move = chess.Move(
            _flip_rank(best_canonical.from_square),
            _flip_rank(best_canonical.to_square),
            promotion=best_canonical.promotion,
        ).uci()
    else:
        best_move = best_canonical.uci()

    return scalar, wdl, moves_left, legal_moves, legal_lookup, policy_probs, best_move
